Marlin/src/module/stepper.h

```diff
@@ -330,35 +330,38 @@ constexpr ena_mask_t enable_overlap[] = {
 
 #if HAS_SHAPING
 
-  // These constexpr are used to calculate the shaping queue buffer sizes
-  constexpr xyze_float_t max_feedrate = DEFAULT_MAX_FEEDRATE;
-  constexpr xyze_float_t steps_per_unit = DEFAULT_AXIS_STEPS_PER_UNIT;
-  // MIN_STEP_ISR_FREQUENCY is known at compile time on AVRs and any reduction in SRAM is welcome
-  #ifdef __AVR__
-    constexpr float max_isr_rate = _MAX(
-                                      LOGICAL_AXIS_LIST(
-                                        max_feedrate.e * steps_per_unit.e,
-                                        max_feedrate.x * steps_per_unit.x,
-                                        max_feedrate.y * steps_per_unit.y,
-                                        max_feedrate.z * steps_per_unit.z,
-                                        max_feedrate.i * steps_per_unit.i,
-                                        max_feedrate.j * steps_per_unit.j,
-                                        max_feedrate.k * steps_per_unit.k,
-                                        max_feedrate.u * steps_per_unit.u,
-                                        max_feedrate.v * steps_per_unit.v,
-                                        max_feedrate.w * steps_per_unit.w
-                                      )
-                                      OPTARG(ADAPTIVE_STEP_SMOOTHING, MIN_STEP_ISR_FREQUENCY)
-                                    );
-    constexpr float max_step_rate = _MIN(max_isr_rate,
-                                      TERN0(INPUT_SHAPING_X, max_feedrate.x * steps_per_unit.x) +
-                                      TERN0(INPUT_SHAPING_Y, max_feedrate.y * steps_per_unit.y)
-                                    );
+  // constexpr for calculating the shaping queue buffer sizes
+  constexpr xyze_float_t DMF = DEFAULT_MAX_FEEDRATE,
+                         SPU = DEFAULT_AXIS_STEPS_PER_UNIT;
+
+  #ifdef SHAPING_MAX_STEPRATE
+    constexpr float max_step_rate = SHAPING_MAX_STEPRATE;
   #else
-    constexpr float max_step_rate = TERN0(INPUT_SHAPING_X, max_feedrate.x * steps_per_unit.x) +
-                                    TERN0(INPUT_SHAPING_Y, max_feedrate.y * steps_per_unit.y);
+    constexpr float max_shaped_rate = TERN0(INPUT_SHAPING_X, DMF.x * SPU.x) +
+                                      TERN0(INPUT_SHAPING_Y, DMF.y * SPU.y);
+    #if defined(__AVR__) || !defined(ADAPTIVE_STEP_SMOOTHING)
+      // MIN_STEP_ISR_FREQUENCY is known at compile time on AVRs and any reduction in SRAM is welcome
+      template<int INDEX=DISTINCT_AXES> constexpr float max_isr_rate() {
```

Comment:
Should this be considering the E axis in trying to figure out the max feedrate? I guess some of our motion code will attempt to up the frequency when possible via `ADAPTIVE_STEP_SMOOTHING`, but how does E come into play when dealing with Input Shaping?

Reply:
Yes, sadly E has to be included.

Entries are added to the shaping buffer whenever bresenham would normally generate a step for any shaped axis. So when X and Y are both shaped that could conceivably be as much as (steprate for X + steprate for Y). However if (steprate for X + steprate for Y) exceeds the rate at which the bresenham logic is called then there is definitely going to be some overlap where X and Y steps are generated at the same time and only one buffer entry is needed. The rate at which the bresenham logic is called is that of the fastest stepping axis (or `MIN_STEP_ISR_FREQUENCY` if ASS is enabled). The fastest stepping axis could be the E axis. So its rate has to be taken into account.